src/views/Activate/ActivateCode.js

```diff
@@ -1,150 +1,166 @@
+import { Grid, H4 } from 'indigo-react';
 
 import View from 'components/View';
 import { ForwardButton } from 'components/Buttons';
 import Passport from 'components/Passport';
 
 import { useHistory } from 'store/history';
 
-import { useTicketInput } from 'lib/useInputs';
 import * as need from 'lib/need';
 import { ROUTE_NAMES } from 'lib/routeNames';
-import { useSyncKnownPoints } from 'lib/useSyncPoints';
 import FooterButton from 'components/FooterButton';
-import { blinkIf } from 'components/Blinky';
 import { DEFAULT_HD_PATH, walletFromMnemonic } from 'lib/wallet';
 import { useNetwork } from 'store/network';
 import { generateWallet } from 'lib/invite';
 import { generateTemporaryOwnershipWallet } from 'lib/walletgen';
 import { useActivateFlow } from './ActivateFlow';
 import { useLocalRouter } from 'lib/LocalRouter';
 import useImpliedTicket from 'lib/useImpliedTicket';
+import timeout from 'lib/timeout';
+import BridgeForm from 'form/BridgeForm';
+import SubmitButton from 'form/SubmitButton';
+import {
+  TicketInput,
+  hasErrors,
+  composeValidator,
+  buildTicketValidator,
+} from 'form/Inputs';
+import FormError from 'form/FormError';
+import { FORM_ERROR } from 'final-form';
 
 export default function ActivateCode() {
   const history = useHistory();
   const { names, push } = useLocalRouter();
   const { contracts } = useNetwork();
   const impliedTicket = useImpliedTicket();
-  const [generalError, setGeneralError] = useState();
-  const [deriving, setDeriving] = useState(false);
+
+  const cachedInviteWallet = useRef(Nothing());
+  const cachedPoint = useRef();
+
   const {
-    derivedWallet,
     setDerivedWallet,
     setInviteWallet,
     derivedPoint,
     setDerivedPoint,
   } = useActivateFlow();
 
-  const [ticketInput, { pass: validTicket, data: ticket }] = useTicketInput({
-    name: 'ticket',
-    label: 'Activation Code',
-    initialValue: impliedTicket || '',
-    autoFocus: true,
-  });
-
   const goToLogin = useCallback(() => history.popAndPush(ROUTE_NAMES.LOGIN), [
     history,
   ]);
+
   const goToPassport = useCallback(() => push(names.PASSPORT), [names, push]);
 
-  const pass = derivedWallet.matchWith({
-    Nothing: () => false,
-    Just: () => true,
-  });
+  // validate should be a pure function but we don't want to have to recompute
+  // all of this information on submit, so cache the invite wallet and avoid
+  // re-renders that may trigger re-validations (causing infinite loop)
+  const validateForm = useCallback(
+    async ({ ticket }, errors) => {
+      if (hasErrors(errors)) {
+        return errors;
+      }
+
+      console.log('FUCK — deriving seed');
+
+      await timeout(100); // allow the ui changes to flush before we lag it out
 
-  useEffect(() => {
-    if (validTicket) {
       const _contracts = need.contracts(contracts);
+      const { seed } = await generateTemporaryOwnershipWallet(ticket);
+
+      // TODO(fang): isn't all this accessible in the ownership object?
```

Comment:
I thought I touched this up already? We can just use `generateResult.keys.address` etc. But Maybe that structure doesn't quite match what's currently in use.

Reply:
this is just a modification of the comment that was already there — cargo-culted it — will remove